src/kmeans.jl

```diff
@@ -1,91 +1,115 @@
 # K-means algorithm
 
-#### Interface
-
-mutable struct KmeansResult{T<:Real} <: ClusteringResult
-    centers::Matrix{T}         # cluster centers (d x k)
-    assignments::Vector{Int}   # assignments (n)
-    costs::Vector{T}           # costs of the resultant assignments (n)
-    counts::Vector{Int}        # number of samples assigned to each cluster (k)
-    cweights::Vector{Float64}  # cluster weights (k)
-    totalcost::Float64         # total cost (i.e. objective) (k)
-    iterations::Int            # number of elapsed iterations
-    converged::Bool            # whether the procedure converged
+####
+#### Result object
+####
+
+struct KmeansResult{TC<:AbstractFloat,TD<:Real,TCW<:Real} <: ClusteringResult
+    centers::Matrix{TC}       # cluster centers (p x k)
+    assignments::Vector{Int}  # assignments (n)
+    costs::Vector{TD}         # cost of the assignments (n)
+    counts::Vector{Int}       # number of samples assigned to each cluster (k)
+    cweights::Vector{TCW}     # cluster weights (k)
+    totalcost::TD             # total cost (i.e. objective)
+    iterations::Int           # number of elapsed iterations
+    converged::Bool           # whether the procedure converged
 end
 
 const _kmeans_default_init = :kmpp
 const _kmeans_default_maxiter = 100
 const _kmeans_default_tol = 1.0e-6
 const _kmeans_default_display = :none
 
-function kmeans!(X::AbstractMatrix{T}, centers::AbstractMatrix{T};
-                 weights::Union{Nothing, AbstractVector{<:Real}}=nothing,
-                 maxiter::Integer=_kmeans_default_maxiter,
-                 tol::Real=_kmeans_default_tol,
-                 display::Symbol=_kmeans_default_display,
-                 distance::SemiMetric=SqEuclidean()) where T<:Real
-
-    m, n = size(X)
-    m2, k = size(centers)
-    m == m2 || throw(DimensionMismatch("Inconsistent array dimensions."))
+####
+#### Exported kmeans and kmeans! functions
+####
+
+function kmeans!(X::AbstractMatrix{TX},                    # in: sample matrix (p x n)
+                 centers::AbstractMatrix{TC};              # in: current centers (p x k)
+                 weights::Union{Nothing, AbstractVector{TW}}=nothing, # in: sample weights (n)
+                 maxiter::Integer=_kmeans_default_maxiter, # in: maximum number of iterations
+                 tol::Real=_kmeans_default_tol,            # in: tolerance of change at convergence
+                 display::Symbol=_kmeans_default_display,  # in: level of display
+                 distance::SemiMetric=SqEuclidean()        # in: function to compute distances
+                 ) where {TX<:Real,TC<:Real,TW<:Real}
```

Comment:
Ok for `TX->T` but there needs to be a reference for the type of `weights`  otherwise there will be things like  `eltype(weights)`  floating around and surely that's less efficient than dispatching over it?

Reply:
Ok so summarising you're in favour of 

1. using `T` for `X` and it being the only explicit parametric type
1. using `eltype(...)` for other things where needed
1. in the result struct, have it be parametrised on a single type (which will be that of centers, not that of `X`) and have the rest (like costs) be declared as `Vector{<:Real}`?